exonum/src/runtime/mailbox/blockchain_secretary.rs

```diff
@@ -0,0 +1,258 @@
+                    fork.rollback();
+
+                    warn!(
+                        "Error occurred during request {:?} within context {:?}: {:?}",
+                        request, self.context, &err
+                    );
+                    self.check_if_should_stop(err)?;
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Depending on the context of the mailbox flush the fork if needed:
+    /// if requests are independent, we should flush the fork after each request,
+    /// so failure of one request won't affect others.
+    fn flush_if_needed(&self, fork: &mut Fork) {
+        if let MailboxContext::OutsideTxExecution = self.context {
+            fork.flush();
+        }
+    }
+
+    /// Depending on the context of the mailbox decides if we should stop processing
+    /// requests or can continue.
+    fn check_if_should_stop(&self, err: ExecutionError) -> Result<(), ExecutionError> {
+        match self.context {
+            MailboxContext::TxExecution(_) => {
+                // Requests created within transaction execution process are considered
+                // highly tied (since multiple requests are possible only if one service
+                // called transaction of other service, and both of them requested an action),
+                // so failure of the one means failure of others too.
+                Err(err)
+            }
+            _ => {
+                // If this is not a transaction execution context (e.g. `before_commit`),
+                // requests are considered independent, so we can safely skip failed one.
+                Ok(())
+            }
+        }
+    }
+
+    fn process_request(
+        &self,
+        mailbox: &mut BlockchainMailbox,
+        dispatcher: &mut Dispatcher,
+        fork: Option<&mut Fork>,
+        request: Action,
+        call_after: AfterRequestCompleted,
+    ) -> Result<(), ExecutionError> {
+        // After the action execution we should call the callback
+        // if it is provided.
+        // Since callback is untrusted code, it's wrapped into
+        // `catch_panic`.
+        let callback = move |()| {
+            catch_panic(move || {
+                if let Some(callback) = call_after {
+                    callback();
+                }
+                Ok(())
+            })
+        };
+
+        catch_panic(|| match request {
+            // Immutable action.
+            Action::StartDeploy { artifact, spec } => {
+                // Request the dispatcher to start deploy process and wait until it completed.
+                // Please note that it doesn't mean completion of the deployment process,
+                // because after that artifact should be registered as deployed.
+
+                // TODO: Get rid of `wait` here.
+                dispatcher.deploy_artifact(artifact, spec).wait()
+            }
+
+            // Mutable action.
```

Comment:
The whole concept of mutable actions (i.e., actions requiring a fork, which, as far as I can see, stems from dispatcher having its own `Schema`) looks like a sign of flawed design.

Reply:
I don't think that we can easily get rid of the dispatcher schema right now.